server/src/main/java/org/elasticsearch/plugins/internal/document_parsing_observer/DocumentParsingObserver.java

```diff
@@ -0,0 +1,41 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.plugins.internal.document_parsing_observer;
+
+import org.elasticsearch.xcontent.XContentParser;
+
+/**
+ * An interface to allow wrapping an XContentParser and observe the events emitted while parsing
+ * A default implementation returns a noop DocumentParsingObserver - does not wrap a XContentParser and
+ * does not do anything upon finishing parsing.
+ */
+public interface DocumentParsingObserver {
+    /**
+     * a default noop implementation
+     */
+    DocumentParsingObserver EMPTY_INSTANCE = new DocumentParsingObserver() {
+    };
+
+    /**
+     * Decorates a provided xContentParser with additional logic (gather some state).
+     * The Decorator parser should use a state from DocumentParsingObserver
+     * in order to allow a parsingFinished method to use that state
+     * @param xContentParser to be decorated
+     * @return a decorator xContentParser
+     */
+    default XContentParser wrapParser(XContentParser xContentParser) {
+        return xContentParser;
+    }
+
+    /**
+     * Reports (perhaps to some external components) the state that was gathered by a decorated wrap
+     * @param indexName an index name that is associated with the parsed document
+     */
+    default void parsingFinished(String indexName) {}
```

Comment:
Do either of these implementations need to be `default`? By implementing this interface, you would be providing implementations of both of these, yes?

Reply:
> re naming, let's just have this as setIndexName and make this class Closeable. close() can indicate parsing is finished.

it wouldn't work. as per my comment here https://github.com/elastic/elasticsearch/pull/97961#issuecomment-1653810303
I would have to add yet another method with a flag to indicate if that the reporting can occur in close (because parsing can be triggered again if dynamic mappings are created). I think this would look strange

Reply:
Yes - the empty_instance provides no-op implementations, but other implementations have to provide methods

Reply:
> the auto -closeable would not really work - it also needs an information if dynamic mappings were created (and not report if so). hence that DocumentObserver/Reporter would also need that information.

How is:

```
observer.parsingFinished(indexName);
```

different from

```
observer.setIndexName(indexName);
observer.close();
```

Reply:
ah I see, I thought you actually meant AutoClosable
I will make it just Closable